pkg/nsx/services/firewall.go

```diff
@@ -80,189 +71,211 @@ func InitializeSecurityPolicy(NSXClient *nsx.Client, cf *config.NSXOperatorConfi
+	switch sp := obj.(type) {
+	case *v1alpha1.SecurityPolicy:
+		var err error
+		nsxSecurityPolicy, nsxGroups, err = service.buildSecurityPolicy(sp)
 		if err != nil {
+			log.Error(err, "failed to build SecurityPolicy")
 			return err
 		}
-		err = service.GroupStore.Delete(group)
+	case types.UID:
+		indexResults, err := service.SecurityPolicyStore.ByIndex(util.TagScopeSecurityPolicyCRUID, string(sp))
 		if err != nil {
+			log.Error(err, "failed to get security policy", "UID", string(sp))
 			return err
 		}
-		log.Info("successfully delete group", "group", group)
-	}
-	return nil
-}
-
-func (service *SecurityPolicyService) deleteGroup(groupsClient domains.GroupsClient, nsxGroup *model.Group) error {
-	failIfSubtreeExistsParam := false
-	forceParam := false
-	return groupsClient.Delete(getDomain(service), *nsxGroup.Id, &failIfSubtreeExistsParam, &forceParam)
-}
+		if len(indexResults) == 0 {
+			log.Info("did not get security policy with index", "UID", string(sp))
+			return nil
+		}
+		t := indexResults[0].(model.SecurityPolicy)
+		nsxSecurityPolicy = &t
 
-func (service *SecurityPolicyService) DeleteSecurityPolicy(UID types.UID) error {
-	groupsClient := service.NSXClient.GroupClient
-	policiesClient := service.NSXClient.SecurityClient
+		indexResults, err = service.GroupStore.ByIndex(util.TagScopeSecurityPolicyCRUID, string(sp))
+		if err != nil {
+			log.Error(err, "failed to get groups", "UID", string(sp))
+			return err
+		}
+		if len(indexResults) == 0 {
+			log.Info("did not get groups with index", "UID", string(sp))
+		}
+		for _, group := range indexResults {
+			*nsxGroups = append(*nsxGroups, group.(model.Group))
+		}
+	}
 
-	policy, err := service.SecurityPolicyStore.ByIndex(util.TagScopeSecurityPolicyCRUID, string(UID))
-	if err != nil {
-		return err
+	nsxSecurityPolicy.MarkedForDelete = &MarkedForDelete
+	for i := len(*nsxGroups) - 1; i >= 0; i-- { // Don't use range, it would copy the element
+		(*nsxGroups)[i].MarkedForDelete = &MarkedForDelete
 	}
-	if len(policy) == 0 {
-		log.V(1).Info("no NSX SecurityPolicy is found", "UID", string(UID))
-		return nil
+	for i := len(nsxSecurityPolicy.Rules) - 1; i >= 0; i-- { // Don't use range, it would copy the element
+		nsxSecurityPolicy.Rules[i].MarkedForDelete = &MarkedForDelete
 	}
-	nsxSecurityPolicy := policy[0].(model.SecurityPolicy)
-	if err = policiesClient.Delete(getDomain(service), *nsxSecurityPolicy.Id); err != nil {
-		return err
+
+	// WrapHighLevelSecurityPolicy will modify the input security policy, so we need to make a copy for the following store update.
+	finalSecurityPolicyCopy := *nsxSecurityPolicy
+	finalSecurityPolicyCopy.Rules = nsxSecurityPolicy.Rules
+	infraSecurityPolicy, error := service.WrapHierarchySecurityPolicy(nsxSecurityPolicy, *nsxGroups)
+	if error != nil {
+		return error
 	}
-	if err = service.SecurityPolicyStore.Delete(nsxSecurityPolicy); err != nil {
+	err := service.NSXClient.InfraClient.Patch(*infraSecurityPolicy, &EnforceRevisionCheckParam)
```

Comment:
Please make the same change to v4.0.0 and v3.2.1 to make this patch consistence with all branches. 
to remove "enforceRevisionCheckParam := false"

Reply:
ok